kubevirt-vm-latency: Spin up VMs

ormergi opened the pull request:
With this PR changes the Kubevirt VM lantecy checkup will spin up two SR-IOV VMs inside its working namespace.
Each VM is configured with custom MAC address and static IP address using cloudinit.

On teadown, is will do best effort to delete the VMs, in case a once of the VMs fail to delete 
or did was not disposed on time it wont fail immediately. 
Instead it will try delete next VM, when finished it will return an error with the failing VMs and errors.

In order to deploy the checkup, apply the manifests under 'checkups/kubevirt-vm-latency/manifests':
from checkups/kubevirt-vm-latency:
```
$ kubectl apply -f ./manifests/sriov-network-attachment-definition.yaml
$ kubectl apply -f ./manifests/clusterroles.yaml
$ kubectl apply -f ./manifests/kubevirt-vm-latency-checkup.yaml
```

View results:
`$ kubectl get configmap kubevirt-vm-latency-checkup-config -n kiagnose -o yaml
`
Cleanup:
```
$ kubectl delete -f ./manifests/kubevirt-vm-latency-checkup.yaml
$ kubectl delete -f ./manifests/clusterroles.yaml
$ kubectl delete -f ./manifests/sriov-network-attachment-definition.yaml
```

Its possible to run the checkup directly without the framework (e.g: for development)
by applying the manifests under 'checkups/kubevirt-vm-latency/manifests/dev':
from checkups/kubevirt-vm-latency:
`$ kubectl apply -f ./manifests/dev/kubevirt-vm-latency-checkup.yaml
`
Ciew results:
`$ kubectl get configmap kubevirt-vm-latency-checkup-results -n kubevirt-vm-latency-checkup -o yaml
`
Clean up:
`
$ kubectl delete -f ./manifests/dev/kubevirt-vm-latency-checkup.yaml
`

This PR does not address bridged networks, currently only SR-IOV networks are supported.

ormergi commented:
Rebased

orelmisan commented on checkups/kubevirt-vm-latency/go.mod:
```diff
@@ -2,7 +2,12 @@ module github.com/kiagnose/kiagnose/checkups/kubevirt-vm-latency
-require github.com/stretchr/testify v1.7.1
+require (
+	github.com/stretchr/testify v1.7.1
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
```
Can you please describe in the commit message what is this package and why do we need it?

ormergi replied:
Following [[1]](https://github.com/kiagnose/kiagnose/pull/55/commits/54012531f84b30928672199a4fbe396da6b0437e#r870963787), now `sigs.k8s.io/yaml` is being used by cloidinit.go, I updated the commit message.

ormergi replied:
I mean for this https://github.com/kiagnose/kiagnose/pull/55#discussion_r870963787

orelmisan commented on checkups/kubevirt-vm-latency/go.mod:
```diff
@@ -2,7 +2,12 @@ module github.com/kiagnose/kiagnose/checkups/kubevirt-vm-latency
+require (
+	github.com/stretchr/testify v1.7.1
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
+	k8s.io/api v0.23.5
+	k8s.io/apimachinery v0.23.5
```
Several lines below you are replacing these to 0.23.1, is it necessary?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/cloudinit.go:
```diff
@@ -0,0 +1,114 @@
+package vmgenerator
```
Can you please explain in the commit message what this file / package does?

EdDev replied:
It will be also polite to specify its source of origin.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/cloudinit.go:
```diff
@@ -0,0 +1,114 @@
+package vmgenerator
```
Can you please add the licence comment?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+package vmgenerator
```
Can you please add the licence comment?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+	kvcorev1 "kubevirt.io/api/core/v1"
+)
+
+func NewLatencyCheckVmi(
+	namespace, name,
+	netAttachDefNamespace, netAttachDefName,
+	macAddress, cidr,
+	nodeName string) *kvcorev1.VirtualMachineInstance {
```
Can you please pass a config struct(s) instead of seven arguments?

EdDev replied:
Or use the build pattern.
I am not really sure why two patterns have been used.

ormergi replied:
> Or use the build pattern.

I didn't picked the builder pattern here since the is only one VM variant and to keep the implementation simple,
you think its better fit in anycase?

> I am not really sure why two patterns have been used

What you mean?

EdDev replied:
> I didn't picked the builder pattern here since the is only one VM variant and to keep the implementation simple

The same reasoning conflicts with the implementation in coudinit, no?

> you think its better fit in anycase?

I think there is a need to keep consistency.
The build pattern has the advantage of reading a bit nicer and resolves the large amount of arguments passed in as a side effect.
Does it merits its usage? I cannot say for sure.

EdDev replied:
> What you mean?

`cloudinit` implementation uses the build pattern, while here you use a more "classic" pattern of mutating an existing object on a procedural manner.

ormergi replied:
Thanks for explaining, I changed it  follow the builder pattern.
Looking at this again its more clear that its worth to follow the builder pattern in this case as it improve readability, it worth the cost of the overhead it brings.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+			Rng: &kvcorev1.Rng{},
+		},
+	}
+	const defaultTerminationGracePeriodSeconds = 5
+	terminationGracePeriodSeconds := int64(defaultTerminationGracePeriodSeconds)
+	vmi.Spec.TerminationGracePeriodSeconds = &terminationGracePeriodSeconds
```
I've seen this trick in Learning Go:
```go
func int64ptr(n int) *int64 {
   return &n
}

vmi.Spec.TerminationGracePeriodSeconds = int64ptr(defaultTerminationGracePeriodSeconds)
```

EdDev replied:
For some reason this looks familiar. See [here](https://github.com/kiagnose/kiagnose/blob/f7184d332626ee159502b43488258d4306e9c0ae/kiagnose/internal/checkup/checkup.go#L217).

orelmisan replied:
> For some reason this looks familiar. See [here](https://github.com/kiagnose/kiagnose/blob/f7184d332626ee159502b43488258d4306e9c0ae/kiagnose/internal/checkup/checkup.go#L217).

I remember this code.
I find the usage of const + a function that returns a pointer to its parameter cleaner.

orelmisan replied:
This should have been `func int64ptr(n int64) *int64 {`.
I wrote it late at night...

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+	vmi.Spec.TerminationGracePeriodSeconds = &terminationGracePeriodSeconds
+
+	addPodNetwork(vmi)
+	addSriovNetwork(vmi, networkName, netAttachDefNamespace+"/"+netAttachDefName, macAddress, cidr)
```
Please consider extracting this part to a function:
```
netAttachDefNamespace+"/"+netAttachDefName
```

EdDev replied:
I would be fine with just `strings.Join(...)`, but I do not mind also a function.

EdDev replied:
Maybe I missed something, but this seems to support SR-IOV only interfaces. I thought we are suppose to support it per the NAD definition.

ormergi replied:
I think I missed it, I thought it should be implemented at another phase of this checkup.

Do you prefer it to be part of this PR? 
if not I will do it on a followup PR.

EdDev replied:
It is fine to do it in a following PR, just please make it clear in this PR that this is the case.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+
+func addPodNetwork(vmi *kvcorev1.VirtualMachineInstance) {
+	vmi.Spec.Networks = append(vmi.Spec.Networks, podNetwork("pod"))
+	vmi.Spec.Domain.Devices.Interfaces = append(vmi.Spec.Domain.Devices.Interfaces, masqueradeInterface("pod"))
```
Why `masqueradeInterface` and not bridge for example?

orelmisan replied:
I was just looking for the reasoning.

This is a better [link](https://github.com/kiagnose/kiagnose/pull/55#discussion_r870968794) to the relevant thread.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+		withEthernet(networkName,
+			withAddresses(cidr),
+			withMatchingMAC(macAddress),
+		))
```
nit: can the second `)` be in a line of its own?

orelmisan reviewed (commented):
Thank you @ormergi for the PR!
Please see the inline comments for the first commit.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -28,13 +28,14 @@ import (
-func New(c *client.Client, params config.CheckupParameters) *checkup {
-	return &checkup{client: c, params: params}
+func New(c *client.Client, workingNamespace string, params config.CheckupParameters) *checkup {
+	return &checkup{client: c, namespace: workingNamespace, params: params}
```
```suggestion
return &checkup{
  client: c, 
  namespace: workingNamespace, 
  params: params
  }
```

ormergi replied:
Done

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/config/config.go:
```diff
@@ -38,6 +38,7 @@ const (
 )
 
 type CheckupParameters struct {
+	WorkingNamespace                     string
```
Why are there two copies of this? (one here and another in the checkup's state)

ormergi replied:
Thanks for the heads up, this is a left over I removed it.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/mainflow.go:
```diff
@@ -38,6 +38,13 @@ func Run(env map[string]string) error {
 	}
 
-	l := launcher.New(checkup.New(c, cfg.CheckupParameters), reporter.New(c, cfg.ResultsConfigMapNamespace, cfg.ResultsConfigMapName))
+	if err != nil {
+		return err
+	}
```
This looks like a leftover.

ormergi replied:
Sorry for the noise, it is a left over, I removed it.

orelmisan reviewed (commented):
Second commit.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -20,29 +20,84 @@
+		params:       params,
+		setupTimeout: defaultSetupTimeout,
+	}
 }
```
Can you please move it above the `SetSetupTimeout` method so it can be the first function?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -20,29 +20,84 @@
+		targetVmiCidr = "192.168.100.20/24"
+	)
+
+	sourceVmi := vmgenerator.NewLatencyCheckVmi(
+		c.namespace, sourceVmiName,
+		c.params.NetworkAttachmentDefinitionNamespace, c.params.NetworkAttachmentDefinitionName,
+		sourceVmiMac, sourceVmiCidr, c.params.SourceNodeName)
```
I've suggested in a previous commit to pass there parameters inside one or two structs (one for the NAD and another for the rest).

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -20,29 +20,84 @@
+	if err := c.client.StartVmis(ctx, sourceVmi, targetVmi); err != nil {
+		return fmt.Errorf("%s: %v", errMessagePrefix, err)
+	}
+	if err := c.client.WaitForVmisReady(ctx, sourceVmi, targetVmi); err != nil {
```
nit: can you please add a newline here?

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+func TestCheckupSetupShouldFailWhen(t *testing.T) {
+	t.Run("failed to create VM", func(t *testing.T) {
```
```suggestion
t.Run("failed to create a VM", func(t *testing.T) {
```

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,75 @@
+func TestCheckupSetupShouldFailWhen(t *testing.T) {
+	t.Run("failed to create VM", func(t *testing.T) {
+		expectedError := errors.New("error test")
+		testCheckup := checkup.New(&clientStub{failStartVmis: expectedError}, "", newTestsCheckupParameters())
```
```suggestion
testCheckup := checkup.New(
  &clientStub{failStartVmis: expectedError },
   "",
    newTestsCheckupParameters(),
 )
```

ormergi replied:
Done

orelmisan replied:
nit: I suggest to put the last `)` on a line of its own.

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,75 @@
+		assert.Equal(t, testCheckup.Setup(), fmt.Errorf("setup: %v", expectedError))
+	})
+
+	t.Run("VMI was not ready on time", func(t *testing.T) {
```
```suggestion
t.Run("VMI was not ready before timeout expiration", func(t *testing.T) {
```

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,75 @@
+
+	t.Run("VMI was not ready on time", func(t *testing.T) {
+		expectedError := errors.New("error test")
+		testCheckup := checkup.New(&clientStub{failWaitForVmisReady: expectedError}, "", newTestsCheckupParameters())
```
Same suggestion as above.

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:
```diff
@@ -55,3 +58,11 @@ func (c *Client) UpdateConfigMap(namespace, name string, date map[string]string)
 	return nil
 }
+
+func (c *Client) StartVmis(ctx context.Context, vmis ...*kvcorev1.VirtualMachineInstance) error {
+	return vmilifecycle.StartVmis(ctx, c.KubevirtClient, vmis...)
+}
+
+func (c *Client) WaitForVmisReady(ctx context.Context, vmis ...*kvcorev1.VirtualMachineInstance) error {
+	return vmilifecycle.WaitForVmisReady(ctx, c, vmis...)
+}
```
Can you please consider moving these methods to the vmi package as functions?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,75 @@
+package checkup_test
```
Missing licence comment.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/client/vmi/vmi.go:
```diff
@@ -0,0 +1,74 @@
+package vmi
```
Missing the licence comment.

orelmisan replied:
It is still missing.

ormergi replied:
Done

orelmisan reviewed (commented):
Third commit.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -34,29 +35,40 @@ import (
 	c.setupTimeout = duration
 }
 
+func (c *checkup) SetTeardownTimeout(duration time.Duration) {
```
```suggestion
func (c *checkup) SetTeardownTimeout(timeout time.Duration) {
```

ormergi replied:
Done

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -107,6 +119,23 @@ func (c *checkup) Run() error {
+	const errMessagePrefix = "teardown"
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.teardownTimeout)
+	defer cancel()
```
Is this a leftover from old usage of another context?

ormergi replied:
It was added before we discussed about using context and timeouts,
I think this should remain at least for the `WaitVmisDispose` which is long blocking operation.
@orelmisan @EdDev WDYT?

orelmisan replied:
I think we should keep it.

ormergi replied:
I changed it so that context is used when waiting for VMs to be ready.

orelmisan replied:
@ormergi, `Teardown` is receiving a context from the outside, why should it have another one?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -107,6 +119,23 @@ func (c *checkup) Run() error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.teardownTimeout)
+	defer cancel()
+
+	var errs []string
```
```suggestion
var errs []error
```

ormergi replied:
This is related to https://github.com/kiagnose/kiagnose/pull/55#discussion_r870687760

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -107,6 +119,23 @@ func (c *checkup) Run() error {
+		errs = append(errs, err.Error())
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("%s: %s", errMessagePrefix, strings.Join(errs, ", "))
+	}
```
Please see PR #52, it has something similar.

ormergi replied:
Note that in #52 at the end `[]string` was expected, here we need to return an error.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -19,7 +19,7 @@ const (
 	testNetAttachDefName      = "blue-net"
 	testSampleDurationSeconds = 1
-	testTimeout               = time.Nanosecond
+	testTimeout               = time.Millisecond
```
Why has this changed here?

ormergi replied:
This is a leftover, sorry for the noise, I removed it.

orelmisan reviewed (commented):
Forth commit.

orelmisan commented:
@ormergi, Can you please add manifest files to:
1. Deploy the relevant ClusterRole (with the required permissions so far).
2. Configure and run this checkup (similar to what we have for the echo checkup).

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/cloudinit.go:
```diff
@@ -0,0 +1,114 @@
+import (
+	"fmt"
+
+	"gopkg.in/yaml.v3"
```
The usage of this package requires `yaml` annotations when defining the structs.
You cloned the file but changed this part, very odd.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+	terminationGracePeriodSeconds := int64(defaultTerminationGracePeriodSeconds)
+	vmi.Spec.TerminationGracePeriodSeconds = &terminationGracePeriodSeconds
+
+	addPodNetwork(vmi)
```
Why do we need the pod network in the VM? If I am not mistaken, we are suppose to test just a single network, i.e. a single vNIC that its connection was defined in the NAD.
We are talking only about the secondary networks.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,162 @@
+	macAddress, cidr,
+	nodeName string) *kvcorev1.VirtualMachineInstance {
+	const (
+		memory                   = "1024Mi"
+		networkName              = "net1"
+		alpineContainerDiskImage = "quay.io/kubevirt/alpine-container-disk-demo:v0.52.0"
+		containerDiskName        = "alpine"
```
I do not understand how this was composed.
- Alpine requires very small memory, in the range of 128M - 256M to work.
- By default, it does not support could-init, but this commit hints it is needed.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/ee0c76fca865b04dfc892ca051ae42349bdcd3bc..f2a806d72941fd69e3a8523fb6ce7c9c914cbe3d): Rebased to latest main in order to get #49 changes

ormergi reviewed (commented):
[Changes](https://github.com/kiagnose/kiagnose/compare/f2a806d72941fd69e3a8523fb6ce7c9c914cbe3d..08dc03bf9844c690690f988d21c04b17f5fb9b2d): Addressed "kubevirt-vm-latency: Introduce vm-generator" review comments

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/08dc03bf9844c690690f988d21c04b17f5fb9b2d..41025dfafe9396a6c287854ce10769c98d508fd3): Fixed commit message and reference link

ormergi reviewed (commented):
[Changes](https://github.com/kiagnose/kiagnose/compare/41025dfafe9396a6c287854ce10769c98d508fd3..4051c0db1040cf735d06b54d88ce3523685d3c74): Addressed "kubevirt-vm-latency: Read namespace file and pass it to checkup
" commit review comments

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,160 @@
+	const (
+		memory                   = "1024Mi"
+		networkName              = "net1"
+		alpineContainerDiskImage = "quay.io/kubevirt/fedora-with-test-tooling-container-disk:v0.53.0"
```
@ormergi Please take a look at [this](https://groups.google.com/g/kubevirt-dev/c/siDqS8hgvFo) post in the kubevirt's mailing list. 
Furthermore, the container disk image is fedora but the const name is alpine.

ormergi replied:
Done, note that this is different image and its maintained.

ormergi replied:
Note that now the image is change to alpine-with-test-tooling.

ormergi reviewed (commented):
[Changes](https://github.com/kiagnose/kiagnose/compare/4051c0db1040cf735d06b54d88ce3523685d3c74..1b84913391a86bfd8a91a8e03c4c59200139c4d5): Address "kubevirt-vm-latency: Create VMs on setup" commit review comments.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/1b84913391a86bfd8a91a8e03c4c59200139c4d5..7cdc79cea926b11e4cd71bd83775582fdbeda482): Move functions from vmi.go to client.go.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/1b84913391a86bfd8a91a8e03c4c59200139c4d5..7cdc79cea926b11e4cd71bd83775582fdbeda482): Move functions from vmi.go to client.go, @orelmisan let me know if this what you meant.

ormergi reviewed (commented):
[Changes](https://github.com/kiagnose/kiagnose/compare/7cdc79cea926b11e4cd71bd83775582fdbeda482..43f2849ab2265f1ad5794370457dd2bf70965926): Addressed "kubevirt-vm-latency: Delete VMs on teardown" commit review comments.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/43f2849ab2265f1ad5794370457dd2bf70965926..dd5ee51a7aa93206bd220514430a7a955d51af24): fix VM image variable name, @orelmisan note that this is different image and its maintained.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/dd5ee51a7aa93206bd220514430a7a955d51af24..f22a0493a1c0c6c71a46081d26ae7527f244ccb8): Move `checkup.New()` closer to the type definition

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/f22a0493a1c0c6c71a46081d26ae7527f244ccb8..0663fd5cf88237fb26d76de2bba8b673f6e6d5f3): revert https://github.com/kiagnose/kiagnose/pull/55#issuecomment-1125008793

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/0663fd5cf88237fb26d76de2bba8b673f6e6d5f3..17d813f5ce237a9aee5183e69146c5a4819a7231): Fix license headers, rename variable names at `vmgenerator`

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/17d813f5ce237a9aee5183e69146c5a4819a7231..a4e3dce6b3761273ca80cb6d0d57dd3fff81cb1f): Set the VM's system to boot first explicitly, it prevents from cloudinit disk to boot first.
it seems that if no boot order is set, the disk will boot the they order in the VM spec.

ormergi commented:
@EdDev @orelmisan I added manifests and readme.
I verified that the manifests are valid and the checkup is working on a local setup, please take another look.

orelmisan commented:
Thank you for the changes @ormergi!
Is it possible to please split the README file to a separate PR?
Unlike the manifests, it is not mandatory for the ability to run the checkup.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/cloudinit.go:
```diff
@@ -0,0 +1,136 @@
+/*
+ *
+ * source: https://github.com/kubevirt/kubevirt/blob/v0.53.0/tests/libnet/cloudinit.go
```
I am not familiar with such an annotation in the license itself. Do you have some ref for other projects doing the same with this license type?
You could document the package itself and then add such a contribution there.

EdDev replied:
Either drop it from here or place it in a package doc, whatever you prefer.
Unless you find a ref about doing this as it is presented in the change here.

ormergi replied:
Thanks for the heads up, I changed it to be the package doc.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,160 @@
+/*
+ * source: https://github.com/kubevirt/kubevirt/blob/v0.53.0/tools/vms-generator/utils/utils.go
```
Same

ormergi replied:
Done

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,160 @@
+	k8scorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kvcorev1 "kubevirt.io/api/core/v1"
```
nit: Please separate imports per domain (with newline).

We need to find a formatter that does this automatically.

ormergi replied:
Done, all over.

I suggest to not be strict about it until there will be such formatter, its very easy to miss it.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -0,0 +1,160 @@
+		memory                   = "1024Mi"
+		networkName              = "net1"
+		fedoraContainerDiskImage = "quay.io/kubevirt/fedora-with-test-tooling-container-disk:v0.53.0"
+		containerDiskName        = "alpine"
```
These are not in sync.
I would also suggest using the new Alpine with tooling, but it is not a must for this PR.

ormergi replied:
It was changed at "kubevirt-vm-latency: Create VMs on setup" commit, I will move it to this commit.

> I would also suggest using the new Alpine with tooling, but it is not a must for this PR.

I wasn't aware there such image, does it support cloudinit?

EdDev replied:
Yes, it originated [here](https://github.com/kubevirt/kubevirtci/pull/770). It is already used in [kubevirt e2e tests](https://github.com/kubevirt/kubevirt/blob/00b1544db6b7ef8ef19e10509544784e0475be7a/tests/libvmi/factory.go#L78).

ormergi replied:
Thanks for the heads-up I changed it to use alpine-with-test-tooling image.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -28,13 +28,18 @@ import (
 
-func New(c *client.Client, params config.CheckupParameters) *checkup {
-	return &checkup{client: c, params: params}
+func New(c *client.Client, workingNamespace string, params config.CheckupParameters) *checkup {
```
nit: In this context, the "working" prefix has no meaning (it has a meaning to the caller maybe). I suggest to drop it.

ormergi replied:
Done

ormergi replied:
I moved the change to the right commit (second commit)

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/mainflow.go:
```diff
@@ -27,7 +27,7 @@ import (
 )
 
-func Run(env map[string]string) error {
+func Run(env map[string]string, workingNamespace string) error {
```
Same, I suggest dropping `working` prefix.

ormergi replied:
Done

EdDev replied:
Sill here

ormergi replied:
I moved the change to the right commit (second commit)

EdDev reviewed (changes requested):
> In K8s each in container there is a file that containes its namespace
by default.

I do not understand this sentence.

EdDev reviewed (commented):
Commit `kubevirt-vm-latency: Read namespace file and pass it to checkup`:

Could you please elaborate in the commit message about:
- The expectation of the executable to be run in a kubernetes pod, where the namespace file is mounted.
- The decision to create VM/s in the namespace from which the checkup is executed. After all, this was not the decision taken from the ConfigMap (it was passed from kiagnose as input).

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/vmgenerator/vmgenerator.go:
```diff
@@ -131,9 +131,13 @@ func newCloudinitVolumeWithNetworkData(name, networkData string) kvcorev1.Volume
 }
 
-func addContainerDiskVolume(vmi *kvcorev1.VirtualMachineInstance, volumeName, image string) {
+func addSystemContainerDiskVolume(vmi *kvcorev1.VirtualMachineInstance, volumeName, image string) {
```
I am not sure I really like this usage of the boot order.
The building blocks used in [libvmi](https://github.com/kubevirt/kubevirt/blob/main/tests/libvmi/storage.go) seems to work fine. Adding disks in the right order can also be done here, so placing this hard definition seems unneeded. Is there some hidden reason behind all this that I missed?

ormergi replied:
While testing this PR on local setup, it seemed that if the boot order is not configured explicitly the disks boot order will be as it is in the VMI spec, the VM did not boot until I changed the order of the disks in the spec
in a way that the system disk is first.

Note that [libvmi](https://github.com/kubevirt/kubevirt/blob/main/tests/libvmi/factory.go#L39) exposes few factories that make sure that the system disk is added first by placing it at the first at options array.
But If the order is changed so that different disk will be added first, the VM wont boot.
In Kubevirt it is covered by e2e tests, as if the order is messed a PR wont pass CI.

In general I think that its wrong to not set boot order explicitly when there is more then one disk in the spec.

Here we have dont have e2e tests just yet, instead of changing the order of lines 63-64 I decided to set the boot order of the system disk explicitly.
So that if the caller "NewLatencyCheckVmi" is mistakenly changes in a way that cloudinit disk is added first, 
the OS disk wont boot first and this checkup breaks.

EdDev replied:
I think I covered this by saying:
> Adding disks in the right order can also be done here

The caller that constructs the VMI spec controls its correct composition. AFAIK we are not developing a generic library such that someone can create the same in a different order.
Everything is under "our" control, and explicitly setting the boot order by using dedicated parameters in the spec has no better outcome than defining/building the spec in the right order. All parts are hard-coded in this application.

I can even reason that imposing the boot order as you did will not fit all options in the future. A future author may think the order is per the ordering in the spec, and be surprised that it is hard-coded in the helpers to a specific one.
But this is not the strongest reasoning for not doing this, the strongest reasoning is just that this is not the minimum needed to accomplish the objective. And I personally think the spec should be minimalistic, as each parameter used in it requires maintenance (e.g. knowing what it is and how it affects the VM).

ormergi replied:
Thanks for explaining, I removed explicit boot order definition for now as part of changing the implementation to follow the builder pattern.

I still think its good practice to set the boot order explicitly as long Kubevirt will set the boot order by default (intentionally or not), it should be considered it as the system-disk that should be booted first.
In wonder if the a VM system disk should be expressed explicitly in the API,
this discussion would be obsolete if there was systemDisk attribute in the VM spec,
it seems that its not a thing.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/client/vmi/vmi.go:
```diff
@@ -0,0 +1,94 @@
+	"kubevirt.io/client-go/kubecli"
+)
+
+func StartVmis(ctx context.Context, c kubecli.KubevirtClient, vmis ...*kvcorev1.VirtualMachineInstance) error {
```
I would prefer you call each from the caller explicitly, having such a wrapper does not give any special abstraction.
Even further, all the helpers that try to handle a collection of VMI/s, I think should not exist.

ormergi replied:
Note that having such helpers (that handle connection of VMI's) reduce noise from the caller (checkup.Run()),
this is why I added them in the first place.
You think its better to have their logic at the caller in anycase?

EdDev replied:
I think that showing that **two** VMI/s have been called is informative and does not create noise.

ormergi replied:
Thanks for explaining, I agree it will surly improve visibility.
I removed the helpers that handle VMs collections.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/client/client.go:
```diff
@@ -50,3 +53,11 @@ func (c *Client) UpdateConfigMap(namespace, name string, date map[string]string)
 	return nil
 }
+
+func (c *Client) StartVmis(ctx context.Context, vmis ...*kvcorev1.VirtualMachineInstance) error {
```
I think this is limiting the client to a very specific use case and makes it harder to evolve it.
The `client` should be a simple proxy to the actual client and not attempt to wrap it with logic (even if it is a simple one). We do not know how will the callers evolve and by doing this logic wrapping it limits what can be done.

I have suggested a way to organize the spec and operation of VMI/s, please consider that one.

ormergi replied:
Thanks for explaining, I agree it less flexible the benefit is that it makes it much easier to test,
and for this checkup I think there will be no need for the client to be flexible.

Having said that I do agree we cant predict how it will evolve, and the the fact that
there is no single package that handles such things.
I changed `client` to  wrap the most basic operations of Kubevirt client.

EdDev reviewed (changes requested):
Commit `kubevirt-vm-latency: Create VMs on setup`:

- The change includes some changes (/fix) to the containerdisk, which is a different topic to the spinning of VM/s. I think it merits its own commit with a proper explanation.
It may also be a reasonable simple amend to the commit that introduced it all in the first tplace.
- The mentioning of SR-IOV is mentioned here explicitly for the first time. Please help the reviewer understand why it is now supporting only that and not the general scenario (I would claim it needs to be mentioned a commit before this one).
- The existence of two packages, `vmigenerator` and `vmi`, seems a bit strange to me.
How is the split of the two helps here? Unless you have a specific reason not to do so, I would suggest to place both under a `vmi` package, with two files: `spec` and `operation`. And you could also consider placing the operations as methods.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -34,32 +35,43 @@ import (
+		namespace:       workingNamespace,
+		params:          params,
+		setupTimeout:    defaultSetupTimeout,
+		teardownTimeout: defaultTeardownTimeout,
```
Why this needs to reside in the object state? It looks to me like it is an internal value taken from a const.
I guess I should have asked the same about `setupTimeout`.

ormergi replied:
It was necessary to add `teardownTimeout` and expose `SetTeardownTimeout()` in order to control it in [tests](https://github.com/kiagnose/kiagnose/pull/55/commits/a4e3dce6b3761273ca80cb6d0d57dd3fff81cb1f#diff-9ab680d21e50243435b2107de32e969fa3a3ace87ac5b42d714af101bdbbc97cR77).
If `teardownTimeout` were locals at `Teardown() `the tests would hang until the timeout is reached.
Same goes for `setupTimeout` and `SetSetupTimeout()`.

EdDev replied:
You could have passed the values as an argument to the respective method.
Or even pass in a context to these methods (which may fit nicer, as it expresses that the caller controls this timeout).

ormergi replied:
Done.

ormergi reviewed (commented):
@EdDev I answered to few comments I wasn't sure about.

ormergi commented:
@EdDev 
> Commit `kubevirt-vm-latency: Create VMs on setup`:
> 
> * The change includes some changes (/fix) to the containerdisk, which is a different topic to the spinning of VM/s. I think it merits its own commit with a proper explanation.
>   It may also be a reasonable simple amend to the commit that introduced it all in the first tplace.

Sorry for this mess I will move the changes to the first commit (that introduce this logic).

> * The mentioning of SR-IOV is mentioned here explicitly for the first time. Please help the reviewer understand why it is now supporting only that and not the general scenario (I would claim it needs to be mentioned a commit before this one).

I missed this one, it should have support the generic case.
Do you mind if I will implement the generic case in a followup PR?

> * The existence of two packages, `vmigenerator` and `vmi`, seems a bit strange to me.
>   How is the split of the two helps here? Unless you have a specific reason not to do so, I would suggest to place both under a `vmi` package, with two files: `spec` and `operation`. And you could also consider placing the operations as methods.

With the current implementation, `client/vmi` includes the "operations" and `checkup/vmigenerator` includes the "spec",
it is splitted that way to reduce noise from the callers at `checkup` (Setup(), Teardown()..).
Putting all your comments together regarding this topic:
- `internal/client` - will wrap the most basic operations of Kubevirt client (create, delete..)
- `internal/vmi/spec.go` - provide factory for the checkup VMs spec.
- `internal/vmi/operation.go` - provide helpers for managing the VMs (start, wait, console, ping..)
Is that what you suggest?

EdDev commented:
> Do you mind if I will implement the generic case in a followup PR?

Sounds good to me.

> With the current implementation, `client/vmi` includes the "operations" and `checkup/vmigenerator` includes the "spec", it is splitted that way to reduce noise from the callers at `checkup` (Setup(), Teardown()..). Putting all your comments together regarding this topic:
> 
>     * `internal/client` - will wrap the most basic operations of Kubevirt client (create, delete, console..)
> 
>     * `internal/vmi/spec.go` - provide factory for the checkup VMs spec.
> 
>     * `internal/vmi/operation.go` - provide helpers for managing the VMs (start, wait, console, ping..)
> 
> 
> Is that what you suggest?

Yes, exactly.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/5a7919f6035fd61a1734be717de834c3077b9f1b..adb6bdaa264d45793507cb4a0f5d66719740e424): Rename vmgenerator `internal/checkup/vmigenerator` with `internal/vmi, and change its implementation to builder pattern.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/adb6bdaa264d45793507cb4a0f5d66719740e424..1c673fc0ee7727def10d345f6ce14ee6c24fa396): update go.mod file, no new files.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/cloudinit.go:
```diff
@@ -0,0 +1,135 @@
+
+func WithAddresses(addresses ...string) networkDataInterfaceOption {
+	return func(networkDataInterface *cloudInitInterface) error {
+		if networkDataInterface.Addresses == nil {
+			networkDataInterface.Addresses = []string{}
+		}
```
I do not think this is needed.

ormergi replied:
It is being used in order to set static IP address to the secondary NIC using cloudinit [[1]](https://github.com/kiagnose/kiagnose/compare/5a7919f6035fd61a1734be717de834c3077b9f1b..adb6bdaa264d45793507cb4a0f5d66719740e424#diff-d60aa090110dbe4421acf2db6a61ce9c88ae82264bd0e66ddc999917378b3c22R135)

EdDev replied:
Again, these 3 lines are not needed.
See [here](https://github.com/kubevirt/kubevirt/pull/7748).

EdDev reviewed (commented):
First commit review

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/1c673fc0ee7727def10d345f6ce14ee6c24fa396..464b935a34ea4a977bd44eceda0dca8c4c96e7f0): 
- Move `internal/client/vmi` code to `internal/vmi`
- Change `client` to wrap the most basic Kubevirt client commands.
- Change `checkup` and `launcher` test following the changes to `client` and `vmi`.

ormergi commented:
Changes:
- [[1]](https://github.com/kiagnose/kiagnose/compare/524cb33a09ecc91047a755bd77182cb5d0ccc5f2..464b935a34ea4a977bd44eceda0dca8c4c96e7f0) revert adding commits instead of editing.
- [[2]] https://github.com/kiagnose/kiagnose/compare/464b935a34ea4a977bd44eceda0dca8c4c96e7f0..764d89656eb4d9503243eace37c436d921762350 get rid of helpers that handle collection of VMs.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/764d89656eb4d9503243eace37c436d921762350..bfd97edbbbf8ffbb541e69f4065054dc68b33c84): Reorder  failing checkup teardown tests.

ormergi commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/spec.go:
```diff
@@ -0,0 +1,221 @@
+
+// Package vmi source:
+// https://github.com/kubevirt/kubevirt/tree/v0.53.0/tests/libvmi
+// https://github.com/kubevirt/kubevirt/blob/v0.53.0/tools/vms-generator/utils/utils.go
```
Yes `libvim` is all that needed, this is a left over sorry for the noise, I will remove it.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/bfd97edbbbf8ffbb541e69f4065054dc68b33c84..8702e13314285c22094335fb4a77b4696e2dc8a1): 
- Remove `cloudinit` `withNamespace()`, instead namespace will be passed on creation of VMs.
- Use Kubevirt API `DiskBusVirtio`
- Fix package `vmi` description
- Update commit messages of first two commits

ormergi commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/spec.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+func newDisk(name string) kvcorev1.Disk {
+	const virtioBus = "virtio"
```
Thanks for the heads up, done.

ormergi commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/spec.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// withNamespace sets the namespace.
+func withNamespace(namespace string) Option {
```
I removed this func and changed the code to pass the namespace on creation.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/8702e13314285c22094335fb4a77b4696e2dc8a1..d3c19c905aae6208179406c872500ef9b4f45c2c): Change checkup.Setup() and checkup.Teardown() to receive a context to control their time out instead of exposing methods.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/d3c19c905aae6208179406c872500ef9b4f45c2c..82ec80b0259b6173aefad02c0e6c5e36532fca84): rename `workingNamespace` to just `namespace` where there is no need for it.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/82ec80b0259b6173aefad02c0e6c5e36532fca84..7e080d983c68805760585799dc51931256cf194a) removed the commit that adds readme, I will add it in a follow up PR.

EdDev reviewed (commented):
First two commits still have open issues.

orelmisan reviewed (commented):
Suggestion for the first commit message:
```
kubevirt-vm-latency: Introduce vmi package
The 'vmi' package provides a factory for generating Kubevirt[1] VMs that will be used by the
checkup.
It is built form Kubevirt's 'libvmi' [2] and 'libnet' [3] packages, stripped down to meet the checkup's needs.

Also go.mod file is updated, no new files were added:
- 'kubevirt.io/api' and 'k8s.io/api' are no longer indirect import since they are
   being used directly.
- 'sigs.k8s.io/yaml' required by vmi/cloudinit.go.
[1] https://kubevirt.io
[2] https://github.com/kubevirt/kubevirt/blob/v0.53.0/tests/libvmi
[3] https://github.com/kubevirt/kubevirt/blob/v0.53.0/tests/libnet

Signed-off-by: Or Mergi <ormergi@redhat.com>
```

orelmisan reviewed (commented):
Regarding the second comment message "kubevirt-vm-latency: Read namespace file and pass it to checkup":

> The framework also supports passing parameters to a checkup by specifying it in
the user ConfigMap.

This sentence is confusing.
The framework passes the configuration to the checkup's Job/Pod as environment variables.
The ConfigMap object is only used to report the results (I don't think the term ConfigMap should be in this commit message).

> Thus Kubevirt VM latency checkup VMs can be created only at the checkup namespace.
Unlike any other parameter the checkup namespace is not provided by the ConfigMap.
It is expected that a checkup be aware it executed in Kubernetes pod, and it
should realize the namespace name it belongs to from the mounted namespace file.

By design, the checkup should run in a K8s cluster and set up objects in its own namespace, allocated to it by the framework.
In order to discover this namespace, we read a file which is available in each K8s Pod.

EdDev reviewed (commented):
> kubevirt-vm-latency: Read namespace file and pass it to checkup

The language used here is presenting kiagnose and not the checkup.
I think it should have come from the context of writing a checkup.
I do not understand how the ability to pass other values has anything to do with all this. It is not absurd to think that this could also been passed to the checkup by kiagnose.

Something like this should be enough:
```
Checkups are assigned a dedicated namespace in which they can operate and create objects in.

The existing implementation of kiagnose does not pass the working namespace as input explicitly (using environment variables). Therefore, we are left to learn the namespace in which we work in from the mounter namespace file (which k8s make available).
```

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/073d534f200c850bfa41702a5c7e10e9d8dd6b13..85972943b033617733261f567217ed43562e791f): 
- Updated `kubevirt-vm-latency: Read namespace file and pass it to checkup` and `kubevirt-vm-latency: Introduce vmi package` according to suggestion.
- Removed redundant conditional slice initialization at `vmi` package, `cloudinit.go` `withAddresses()`.
- Removed redundant function description for `vmi` package  `spec.go` `withContainerDiskImage()`

ormergi commented:
@EdDev @orelmisan thanks for the reviews and feedback, I addressed your comments and pushed new changes.
Please take another look :slightly_smiling_face:

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,95 @@
+			testNamespace,
+			newTestsCheckupParameters())
+
+		testCtx, cancel := context.WithTimeout(context.Background(), testTimeout)
```
I do not understand why you need a context with timeout. Isn't he failure occurring before the waiting loop?

ormergi replied:
@EdDev Thanks for the heads up, I removed it.

EdDev commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/operation.go:
```diff
@@ -0,0 +1,78 @@
+	conditionFn := func(ctx context.Context) (bool, error) {
+		updatedVmi, err := c.GetVirtualMachineInstance(namespace, name)
+		if err != nil {
+			return false, nil
```
I do not think it is correct.
Failure to fetch the VMI means there is a problem and there is nothing to retry.`

Once the object is created on the API server, the resource is suppose to be available. If it is not there, something bad happened. If there is a communications error or some intermediate other error, I am not interested to handle it at this level (a controller would probably retry at a higher layer).

But maybe I miss something. Please share the scenario where it makes sense to retry when this fails.

ormergi replied:
@EdDev  Thanks for explaining it make sense,  the current implantation dont breaks anything but I do want to fix it.
Since it will require changes tests, is it ok if I address it in a follow up PR?

EdDev reviewed (commented):
The last commit message needs some more wording, allowing the reader to understand which manifest is required for registration, which one is for running an example job and the purpose of the "dev" ones (and how to use them).

The message also has some typos and wrong folders (e.g. the `manifest` is not seen in the paths).

orelmisan commented:
@ormergi There is a typo in the first commit message:
> It is built form Kubevirt's

form => from

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/spec.go:
```diff
@@ -0,0 +1,209 @@
+
+import (
+	k8scorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
```
nit: Can you please put a new line here?

ormergi replied:
Note that all of them are from the same domain (k8s).

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/vmi/spec.go:
```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+func NewAlpine(name string, opts ...Option) *kvcorev1.VirtualMachineInstance {
```
If this is the main function we are using to instantiate a VM, can it be on top?

ormergi replied:
I think this is a nit, can we address it later?

orelmisan commented on checkups/kubevirt-vm-latency/cmd/environment.go:
```diff
@@ -37,3 +40,14 @@ func envToMap(rawEnv []string) map[string]string {
+	if err != nil {
+		return "", err
+	}
+
```
nit: Please consider adding a check that the file's content is not empty.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -46,16 +54,94 @@ func (c *checkup) Preflight() error {
+	const (
+		errMessagePrefix = "setup"
+
+		defaultSetupTimeout = time.Minute * 10
```
Isn't this timeout very long for an Alpine machine?

ormergi replied:
I set the timeout that long to overcome connection issues or slowness.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -46,16 +54,94 @@ func (c *checkup) Preflight() error {
+			vmi.WithMatchingMAC(macAddress),
+		),
+	)
+	return vmi.NewAlpine(name,
```
nit: Can you please add a new line here?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -46,16 +54,94 @@ func (c *checkup) Preflight() error {
 }
 
-func (c *checkup) Teardown() error {
+func (c *checkup) Teardown(_ context.Context) error {
```
Why was the context added in this commit?

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -46,16 +54,94 @@ func (c *checkup) Preflight() error {
+func newLatencyCheckVmi(
+	name,
+	nodeName,
+	networkName string, netAttachDef types.NamespacedName,
```
What is the difference between `netAttachDef` and `networkName`?

ormergi replied:
`netAttachDef` - namespaced name of the network-attachment-definition object
`networkName` - the name of the interface and network in the VMI spec.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup.go:
```diff
@@ -46,16 +54,94 @@ func (c *checkup) Preflight() error {
+		return fmt.Errorf("%s: %v", errMessagePrefix, err)
+	}
+
+	waitCtx, cancel := context.WithTimeout(ctx, defaultSetupTimeout)
```
From the tests, I understand that you want to control this context from the outside.
So why are there two contexts?

ormergi replied:
In order to set a default timeout.
The parent context is the one that passed to `Setup() `and  controlled by the caller (i, e: tests) 
which may set shorter timeout and cause the context to be canceled before the Setup() default timeout.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -0,0 +1,95 @@
+			testNamespace,
+			newTestsCheckupParameters())
+
+		testCtx, cancel := context.WithTimeout(context.Background(), testTimeout)
```
Here you have a very short timeout and its confusing, because it is not part of the test's subject.

ormergi replied:
Thanks for the heads up, I removed it.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -70,6 +72,58 @@ func TestCheckupSetupShouldFailWhen(t *testing.T) {
+			testClient,
+			testNamespace,
+			newTestsCheckupParameters())
+		testCtx, cancel := context.WithTimeout(context.Background(), testTimeout)
```
Same comment about short timeout in a different scenario.

orelmisan commented on checkups/kubevirt-vm-latency/vmlatency/internal/checkup/checkup_test.go:
```diff
@@ -70,6 +72,58 @@ func TestCheckupSetupShouldFailWhen(t *testing.T) {
+			testClient,
+			testNamespace,
+			newTestsCheckupParameters())
+		testCtx, cancel := context.WithTimeout(context.Background(), testTimeout)
```
AFAIU, the short timeout here should fail the setup.

ormergi replied:
Note that just before Setup() is called, we inject a VMI object in order that makes Setup() pass.

Now I realize there is no need to pass context with short timeout to Setup() in this test, I will remove it.

orelmisan reviewed (commented):
@ormergi Thanks for the changes.
Please see the inline comments.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/d7eeebb6a57de59e37e885a0ff0f0238d52ae81c..33698226cda8f360a15ea486a4aa7bf17565191d): Update commit messages

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/33698226cda8f360a15ea486a4aa7bf17565191d..fd4f9c79bcf61e9a059011ba64b758d955124a1b): Fix commit messages.

ormergi commented on checkups/kubevirt-vm-latency/vmlatency/internal/launcher/launcher_test.go:
```diff
@@ -182,14 +174,45 @@ func (r reporterStub) Report(_ status.Status) error {
+// newFakeClient returns fakeClient that tracks VMIs.
+// The VMI tracker acts as the cluster DB and keep records regarding
+// the VMIs that were created or deleted.
+func newFakeClient() *fakeClient {
```
@EdDev is it ok  with you, if I will get read of using timeouts in a follow up PR?

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/fd4f9c79bcf61e9a059011ba64b758d955124a1b..1790aa56f52a9a84d35b198abb322d992568977d): Remove redandent passing context with timeout in tests.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/1790aa56f52a9a84d35b198abb322d992568977d..5517ee62ae6cdba38d89952b44855cae0e0369da): Fixed second commit message.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/5517ee62ae6cdba38d89952b44855cae0e0369da..c9e83eb6f930e0ab1cbd86908b3084c102a7eaad): Fix tests that hanged due to setting up expected VMI object with the wrong condition.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/c9e83eb6f930e0ab1cbd86908b3084c102a7eaad..447e57ac6696ac3d5841cc3dc0dbb4d7a66d33ba): Change checkup test "Ms were not disposed before timeout expiration" to not pass context with timeout to Setup(), it is redundant.

orelmisan reviewed (approved):
Thank you @ormergi!
